pkg/operation/garden/bootstrap.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package garden
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/discovery"
+	kubernetesclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+	"github.com/gardener/gardener/pkg/utils/version"
+)
+
+// Bootstrapper is a runnable for bootstrapping the garden cluster.
+type Bootstrapper struct {
+	Log        logr.Logger
+	Client     client.Client
+	RESTConfig *rest.Config
+}
+
+// Start runs a soon as the manager got leader.
+func (b *Bootstrapper) Start(parentCtx context.Context) error {
+	// Other controllers depend on garden cluster bootstrapping.
+	// Hence, if we can't bootstrap the garden cluster in a short timeout, terminate and try again after restart.
+	ctx, cancel := context.WithTimeout(parentCtx, 30*time.Second)
+	defer cancel()
+
+	kubernetesClient, err := kubernetesclientset.NewForConfig(b.RESTConfig)
+	if err != nil {
+		return fmt.Errorf("failed creating kubernetes client: %w", err)
+	}
+
+	secretsManager, err := secretsmanager.New(ctx, b.Log.WithName("secretsmanager"), clock.RealClock{}, b.Client, v1beta1constants.GardenNamespace, v1beta1constants.SecretManagerIdentityControllerManager, secretsmanager.Config{})
+	if err != nil {
+		return fmt.Errorf("failed creating new secrets manager: %w", err)
+	}
+
+	if err := bootstrapCluster(ctx, b.Client, kubernetesClient.Discovery(), secretsManager); err != nil {
+		return fmt.Errorf("failed bootstrapping garden cluster: %w", err)
+	}
+
+	if err := secretsManager.Cleanup(ctx); err != nil {
+		return err
```

Comment:
Nit: For symmetry reasons, do you want to provide a more meaningful error message here as well?